# Bike Sharing Analytics

## Task

- Regression: 
		Predication of bike rental count hourly or daily based on the environmental and seasonal settings.
	
- Event and Anomaly Detection: Count of rented bikes are also correlated to some events in the town which easily are traceable via search engines. For instance, query like "2012-10-30 washington d.c." in Google returns related results to Hurricane Sandy. Some of the important events are identified in [1]. Therefore the data can be used for validation of anomaly or event detection algorithms as well.

[1] Fanaee-T, Hadi, and Gama, Joao, "Event labeling combining ensemble detectors and background knowledge", Progress in Artificial Intelligence (2013): pp. 1-15, Springer Berlin Heidelberg, doi:10.1007/s13748-013-0040-3.


## Library

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore") 

## Data Wrangling

### Gathering Data

In [2]:
# Load day data
day_df = pd.read_csv('./data/day.csv')
day_df.head()

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1        0        6           0   
1        2  2011-01-02       1   0     1        0        0           0   
2        3  2011-01-03       1   0     1        0        1           1   
3        4  2011-01-04       1   0     1        0        2           1   
4        5  2011-01-05       1   0     1        0        3           1   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

In [9]:
day_df.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

The dataset contains various columns, including:  
- **instant**: Record index.  
- **dteday**: Date of the record.  
- **season**: The season in which the data was recorded (1: Spring, 2: Summer, 3: Fall, 4: Winter).  
- **yr**: Year of the data (0: 2011, 1: 2012).  
- **mnth**: Month of the year (1 to 12).  
- **holiday**: Whether the day is a holiday (1: Yes, 0: No).  
- **weekday**: Day of the week (0: Sunday, 6: Saturday).  
- **workingday**: Whether the day is a working day (1: Yes, 0: No).  
- **weathersit**: Weather situation (1: Clear, 2: Mist, 3: Light rain/snow, 4: Heavy rain/snow).  
- **temp** : Normalized temperature in Celsius. The values are divided to 41 (max).
- **atemp**: Normalized feeling temperature in Celsius. The values are divided to 50 (max).
- **hum**: Normalized humidity. The values are divided to 100 (max).
- **windspeed**: Normalized wind speed. The values are divided to 67 (max).
- **casual**: Count of casual users.
- **registered**: Count of registered users.
- **cnt**: Count of total rental bikes including both casual and registered.

By examining the first few rows, we can get an overview of the dataset before performing further analysis.  

In [3]:
# Load hour data
hour_df = pd.read_csv('./data/hour.csv')
hour_df.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1

 This dataset contains similar features to the daily dataset (`day_df`), but with hourly records instead of daily aggregates. Some key columns include:  
- **instant**: Unique identifier for each row.  
- **dteday**: Date of the record.  
- **hr**: Hour of the day (0 to 23).  
- **season, yr, mnth, holiday, weekday, workingday, weathersit**: Similar to the daily dataset, representing seasonal and time-related information.  
- **temp, atemp, hum, windspeed**: Weather-related features.  
- **casual, registered, cnt**: Number of casual users, registered users, and total bike rentals for each hour.  

Since this dataset provides a more granular view with hourly records, it allows for a deeper analysis of bike-sharing trends throughout the day.  

### Assessing Data

#### Day data

In [5]:
# Day data information
day_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB



- The dataset contains **731 rows** and **16 columns**.  
- All columns have **731 non-null values**, meaning there are **no missing values**.  
- The dataset consists of different data types:  
  - **Integer (`int64`)**: Used for categorical and count-based features such as `season`, `yr`, `mnth`, `holiday`, `weekday`, `workingday`, `weathersit`, `casual`, `registered`, and `cnt`.  
  - **Float (`float64`)**: Used for continuous numerical values like `temp`, `atemp`, `hum`, and `windspeed`.  
  - **Object (`object`)**: The `dteday` column, which represents dates in string format.  

This information helps in preprocessing and deciding if any columns need transformation (e.g., converting `dteday` to a proper datetime format).  

In [6]:
# Check Duplicate the day data
day_df.duplicated().sum()

0

The output is **0**, which means there are no duplicate records in the dataset. This confirms that each row represents a unique entry, and no additional cleaning is required for duplicates.  

In [7]:
# Check Missing Value the day data
day_df.isna().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

The output shows that **all columns have 0 missing values**, meaning the dataset is complete and does not require imputation or removal of missing data.  

Since there are no missing values, we can proceed with the analysis without additional preprocessing for missing data.  

In [8]:
# The descriptive statistic from day data
day_df.describe()

instant      season          yr        mnth     holiday     weekday  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean   366.000000    2.496580    0.500684    6.519836    0.028728    2.997264   
std    211.165812    1.110807    0.500342    3.451913    0.167155    2.004787   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.500000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    366.000000    3.000000    1.000000    7.000000    0.000000    3.000000   
75%    548.500000    3.000000    1.000000   10.000000    0.000000    5.000000   
max    731.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean     0.683995    1.395349    0.495385    0.474354    0.627894    0.190486   
std      0.465233    0.544894    0.183051    0.162961    0.142429    0.077498   
min      0.000000    1.000000    0.059130    0.079070    0.000000    0.022392   
25%      0.000000    1.000000    0.337083    0.337842    0.520000    0.134950   
50%      1.000000    1.000000    0.498333    0.486733    0.626667    0.180975   
75%      1.000000    2.000000    0.655417    0.608602    0.730209    0.233214   
max      1.000000    3.000000    0.861667    0.840896    0.972500    0.507463   

            casual   registered          cnt  
count   731.000000   731.000000   731.000000  
mean    848.176471  3656.172367  4504.348837  
std     686.622488  1560.256377  1937.211452  
min       2.000000    20.000000    22.000000  
25%     315.500000  2497.000000  3152.000000  
50%     713.000000  3662.000000  4548.000000  
75%    1096.000000  4776.500000  5956.000000  
max    3410.000000  6946.000000  8714.000000

To understand the overall distribution of numerical variables, we use the `.describe()` function. This provides key statistical summaries, including:  
- **count**: Number of non-null values in each column.  
- **mean**: Average value of each column.  
- **std (standard deviation)**: Measures data dispersion from the mean.  
- **min and max**: Minimum and maximum values.  
- **25%, 50% (median), and 75%**: Quartiles, showing data distribution.  

Key Observations:  
- The dataset consists of **731 records**.  
- The **average number of total bike rentals (`cnt`) is 4504**, with a **minimum of 22** and a **maximum of 8714**.  
- The **temperature (`temp`) ranges from 0.06 to 0.86** (normalized values).  
- The dataset is **balanced across years (`yr`)**, with values of `0` (2011) and `1` (2012).  
- The **working day (`workingday`) and holiday (`holiday`) columns are binary**, indicating whether a day is a working day or holiday.  

This summary helps in understanding the overall trends and potential data distributions before further analysis.  

#### Hour data

In [10]:
# Hour data information
hour_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


To understand the structure of the hourly dataset, we use the `.info()` function. This provides an overview of the dataset, including:  
- **Total entries**: 17,379 rows.  
- **Total columns**: 17 attributes.  
- **Data types**:  
  - `int64`: 12 columns (e.g., `season`, `yr`, `mnth`, `hr`, `holiday`, etc.).  
  - `float64`: 4 columns (`temp`, `atemp`, `hum`, `windspeed`).  
  - `object`: 1 column (`dteday`, representing dates).  
- **Memory usage**: 2.3+ MB.  

Key Observations:  
- The dataset contains **hourly records of bike rentals** over two years.  
- There are **no missing values** in any column, which simplifies data preprocessing.  
- The `hr` column represents hours of the day (0-23), allowing for time-based analysis.  
- `cnt` is the total number of rentals, obtained from the sum of `casual` and `registered` users.  

This summary helps in understanding the dataset before performing further exploratory data analysis.  


In [12]:
# Check duplicate the hour data
hour_df.duplicated().sum()

0

The output is **0**, meaning there are **no duplicate entries** in the dataset. This confirms that each row represents a unique hourly record without redundancy.  

In [14]:
# Check Missing Value the hour data
hour_df.isnull().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

The output shows **0 missing values** in all 17 columns. This means the dataset is **fully complete**, with no need for imputation or data cleaning related to missing values.  


In [21]:
# The descriptive statistic from day data
hour_df.describe().round(2)

instant    season       yr      mnth        hr   holiday   weekday  \
count  17379.00  17379.00  17379.0  17379.00  17379.00  17379.00  17379.00   
mean    8690.00      2.50      0.5      6.54     11.55      0.03      3.00   
std     5017.03      1.11      0.5      3.44      6.91      0.17      2.01   
min        1.00      1.00      0.0      1.00      0.00      0.00      0.00   
25%     4345.50      2.00      0.0      4.00      6.00      0.00      1.00   
50%     8690.00      3.00      1.0      7.00     12.00      0.00      3.00   
75%    13034.50      3.00      1.0     10.00     18.00      0.00      5.00   
max    17379.00      4.00      1.0     12.00     23.00      1.00      6.00   

       workingday  weathersit      temp     atemp       hum  windspeed  \
count    17379.00    17379.00  17379.00  17379.00  17379.00   17379.00   
mean         0.68        1.43      0.50      0.48      0.63       0.19   
std          0.47        0.64      0.19      0.17      0.19       0.12   
min          0.00        1.00      0.02      0.00      0.00       0.00   
25%          0.00        1.00      0.34      0.33      0.48       0.10   
50%          1.00        1.00      0.50      0.48      0.63       0.19   
75%          1.00        2.00      0.66      0.62      0.78       0.25   
max          1.00        4.00      1.00      1.00      1.00       0.85   

         casual  registered       cnt  
count  17379.00    17379.00  17379.00  
mean      35.68      153.79    189.46  
std       49.31      151.36    181.39  
min        0.00        0.00      1.00  
25%        4.00       34.00     40.00  
50%       17.00      115.00    142.00  
75%       48.00      220.00    281.00  
max      367.00      886.00    977.00

To understand the distribution of numerical features, we use the `.describe().round(2)` function. This provides key statistical summaries, including count, mean, standard deviation, min, max, and quartiles.  

Key Observations:  
- **Total Entries**: 17,379 rows.  
- **Hourly Distribution** (`hr`): The values range from **0 to 23**, covering all hours in a day.  
- **Temperature (`temp`) & Feels-Like Temperature (`atemp`)**:  
  - `temp` (Normalized between 0 and 1) has a **mean of 0.50**, suggesting an even spread over different weather conditions.  
  - `atemp` follows a similar distribution with a mean of **0.48**.  
- **Humidity (`hum`) & Wind Speed (`windspeed`)**:  
  - Humidity ranges from **0 to 1**, with a mean of **0.63**.  
  - Wind speed has an average of **0.19**, indicating most days experience mild winds.  
- **Bike Rentals (`cnt`)**:  
  - The total count (`cnt`) ranges from **1 to 977** per hour.  
  - Median rentals per hour is **142**, while the upper quartile reaches **281**, indicating some peak hours with high demand.  
  - `casual` riders (mean **35.68**) are significantly fewer than `registered` riders (mean **153.79**), suggesting that most users are registered members.  

Conclusion:  
This statistical summary provides insight into the distribution and variability of bike rentals across different hours, weather conditions, and user types. These insights will guide further analysis and model building.  

### Cleaning Data

In [ ]:
# Drop column when is not using
day_df.drop(['workingday'], axis=1, inplace=True)
hour_df.drop(['workingday'], axis=1, inplace=True)

In [ ]:
# Change data type int to categorical 
columns = ['season', 'mnth', 'holiday', 'weekday', 'weathersit']
for column in columns:
    day_df[column] = day_df[column].astype('category')
    hour_df[column] = hour_df[column].astype('category')

In [ ]:
# Change data type object to datetime
day_df['dteday'] = pd.to_datetime(day_df['dteday'])
hour_df['dteday'] = pd.to_datetime(hour_df['dteday'])

In [ ]:
# Check Again
print(f'df_day["dteday"]: {day_df['dteday'].dtypes}')
print(f'df_hour["dteday"]: {hour_df['dteday'].dtypes}')

In [ ]:
# Change column name from day data
day_df.rename(columns={
    'yr':'year',
    'mnth':'month',
    'weekday':'one_of_week',
    'weathersit':'weather_situation',
    'windspeed':'wind_speed',
    'cnt':'count_cr',
    'hum':'humidity'
},inplace=True)

In [ ]:
# Change column name from hour data

hour_df.rename(columns={
    'yr':'year',
    'hr':'hours',
    'mnth':'month',
    'weekday':'one_of_week', 
    'weathersit':'weather_situation',
    'windspeed':'wind_speed',
    'cnt':'count_cr',
    'hum':'humidity'
},inplace=True)


In [ ]:
# Convert season to be categorical
season_mapping = {
    1: 'Spring', 
    2: 'Summer', 
    3: 'Fall', 
    4: 'Winter'
}

day_df['season'] = day_df['season'].map(season_mapping)
hour_df['season'] = hour_df['season'].map(season_mapping)


In [ ]:
# Convert month to be categorical
month_mapping = {
    1: 'Jan', 2: 'Feb', 3: 'Mar', 4: 'Apr', 5: 'May', 6: 'Jun',
    7: 'Jul', 8: 'Aug', 9: 'Sep', 10: 'Oct', 11: 'Nov', 12: 'Dec'
}
day_df['month'] = day_df['month'].map(month_mapping)
hour_df['month'] = hour_df['month'].map(month_mapping)

In [ ]:
# Convert weather_situation to be categorical 
weather_situation_mapping = {
    1:'Clear', 
    2:'Misty', 
    3:'Light_RainSnow', 
    4:'Heavy_RainSnow'
}
day_df['weather_situation'] = day_df['weather_situation'].map(weather_situation_mapping)
hour_df['weather_situation'] = hour_df['weather_situation'].map(weather_situation_mapping)

In [ ]:
# Convert one_of_week to be categorical
one_of_week_mapping = {
    0: 'Sunday', 
    1: 'Monday', 
    2: 'Tuesday', 
    3: 'Wednesday', 
    4: 'Thursday', 
    5: 'Friday', 
    6: 'Saturday'
}

day_df['one_of_week'] = day_df['one_of_week'].map(one_of_week_mapping)
hour_df['one_of_week'] = hour_df['one_of_week'].map(one_of_week_mapping)

In [ ]:
# Convert year 
year_mapping = {
    0: 2011,
    1: 2012
}
hour_df['year'] = hour_df['year'].map(year_mapping)
day_df['year'] = day_df['year'].map(year_mapping)

In [ ]:
# Count Humidity
day_df['humidity'] = day_df['humidity']*100
hour_df['humidity'] = hour_df['humidity']*100

Make a new column with name days_category when show a column value weekend or weekdays

In [ ]:
# Membuat kolom baru bernama category_days yang menunjukan isi kolom tersebut weekend atau weekdays
one_of_week = [
    "Monday", 
    "Tuesday", 
    "Wednesday", 
    "Thursday", 
    "Friday", 
    "Saturday", 
    "Sunday"
]
def category_days(one_of_week):
    if one_of_week in ["Saturday", "Sunday"]:
        return "weekend"
    else: 
        return "weekdays"

hour_df["category_days"] = hour_df["one_of_week"].apply(category_days)
day_df["category_days"] = day_df["one_of_week"].apply(category_days)

In [ ]:
def classify_humadity(humadity):
    if humadity < 45:
        return "Dry"
    elif humadity >= 45 and humadity < 65:
        return "Normal"
    else:
        return "Humid"
hour_df['humidity_category'] = hour_df['humidity'].apply(classify_humadity) 
day_df['humidity_category'] = day_df['humidity'].apply(classify_humadity)

In [ ]:
# Check again
hour_df.head()

In [ ]:
day_df.head()

## Exploratory Data Analysis (EDA)

### Explore Hour Data

In [ ]:
hour_df.describe(include='all').round(2)

In [ ]:
# Rental by hour
penyewaan_by_jam = hour_df.groupby(by="hours").agg({
    "count_cr": ["sum"]
})

penyewaan_by_jam.head()

### Explore Day Data

In [ ]:
day_df.describe(include='all').round(2)

In [ ]:
# Show total of rental by season
day_df.groupby(by="season").count_cr.sum().sort_values(ascending=False).reset_index()


In [ ]:
# show total rental of each year based on registered dan casual 
day_df.groupby(by="year").agg({
    "registered": ["sum"],
    "casual": ["sum"]
})

##### 

## Visualization & Explanatory Analysis

### Question 1: What are the most and least frequently rented hours?

In [ ]:
sum_order = hour_df.groupby("hours").count_cr.sum().sort_values(ascending=False).reset_index()

fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(30, 15))
sns.barplot(
    x="hours", 
    y="count_cr", 
    data=sum_order.head(5), 
    ax=ax[0],
    color='#ff9999'
)

ax[0].set_ylabel(None)
ax[0].set_xlabel("Hours (PM)", fontsize=30)
ax[0].set_title("Hours with lots of bike renters", loc="center", fontsize=30)
ax[0].tick_params(axis='y', labelsize=35)
ax[0].tick_params(axis='x', labelsize=30)

sns.barplot(
    x="hours", 
    y="count_cr", 
    data=sum_order.sort_values(
        by="hours", 
        ascending=True
    ).head(5), 
    ax=ax[1],
    color='#ff9999'
)

ax[1].set_ylabel(None)
ax[1].set_xlabel("Hours (AM)",  fontsize=30)
ax[1].set_title("Hours with few bike renters", loc="center", fontsize=30)
ax[1].invert_xaxis()
ax[1].yaxis.set_label_position("right")
ax[1].yaxis.tick_right()
ax[1].tick_params(axis='y', labelsize=35)
ax[1].tick_params(axis='x', labelsize=30)

plt.show()


### Question 2: In what season is bike rental peak?

In [ ]:
fig, ax = plt.subplots(figsize=(10, 5))

sns.barplot(
    y="count_cr",
    x="season",
    data= day_df.sort_values(
        by="season",
        ascending=False
    ),
    ax=ax,
    errorbar=None,
    color='#ff9999'
)

ax.set_title("Inter-seasonal Chart", loc="center", fontsize=16)
ax.set_ylabel(None)
ax.set_xlabel(None)
ax.tick_params(axis='x', labelsize=12)
ax.tick_params(axis='y', labelsize=12)

plt.show()

### Question 3: How has the company's sales performance been in recent years?

In [ ]:
plt.figure(figsize=(25,5))

monthly_ct = day_df['count_cr'].groupby(day_df['dteday']).max()

plt.scatter(
    monthly_ct.index,
    monthly_ct.values,
    marker='o',
    color='#ff9999',
    s=10
)
plt.plot(monthly_ct.index,monthly_ct.values)
plt.xlabel('Bulan')
plt.ylabel('Number')
plt.title('Number of Customers per Month in 2012')

plt.show()

### Question 4: How do registered customers compare to casual customers?

In [ ]:
casual_total = sum(day_df['casual'])
total_regiter = sum(day_df['registered'])
data = [total_regiter, casual_total]
labels = ["Registered", "Casual"]
myexplode = [0, 0.1]
colors = [ '#66b3ff', '#ff9999']

plt.pie(data, labels=labels, autopct='%1.1f%%', colors=colors, explode=myexplode)
plt.show()

## Conclusion

- **Question 1:** What are the most and least frequently rented hours?
- **Question 2:** In what season is bike rental peak?
- **Question 3:** How has the company's sales performance been in recent years?
- **Question 4:** How do registered customers compare to casual customers?

**Conclution pertanyaan 1:** Based on the image above, you can see that bike rentals are most used at 17:00. In contrast, rentals at 04:00 are the least sold products.

**conclution pertanyaan 2:** in the Fall season.

**conclution pertanyaan 3:** 
Based on the visualization above, we can see that the highest number of orders occured in September 2012,
we can also see significant decrease in the number of orders in January 2011

**conclution pertanyaan 4:** Those who have registered 81.2%, while those who have not registered (casual) are 18.8% 